lib/mix/lib/mix/dep/loader.ex

```diff
@@ -320,6 +322,13 @@ defmodule Mix.Dep.Loader do
     end
   end
 
+  defp incompatible_requirement?(%Mix.Dep{opts: opts}) do
+    case opts[:lock] do
+      {scm, _, _, _} -> is_nil(opts[scm])
```

Comment:
Thank you! However, this is not the correct fix because the lock entry is not required to be in the shape you are matching on. In other words, this would work only for git and not be generally applicable. I need to investigate further how the proper solution would look like. :)